core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala

```diff
@@ -0,0 +1,224 @@
+
+  /** Create new tokens for the current user and distribute them to the driver. */
+  protected def createAndUpdateTokens(): Unit = {
+    val driver = driverRef.get()
+    require(driver != null, "Driver endpoint not set.")
+
+    val creds = new Credentials()
+    obtainDelegationTokens(creds)
+    UserGroupInformation.getCurrentUser.addCredentials(creds)
+
+    val tokens = SparkHadoopUtil.get.serialize(creds)
+    driver.send(UpdateDelegationTokens(tokens))
+  }
+
+  /**
+   * Fetch new delegation tokens for configured services, storing them in the given credentials.
+   *
+   * @return The time by which the tokens must be renewed.
+   */
+  protected def obtainDelegationTokens(creds: Credentials): Long
+
+  private def scheduleRenewal(delay: Long): Unit = {
+    val _delay = math.max(0, delay)
+    logInfo(s"Scheduling login from keytab in ${UIUtils.formatDuration(delay)}.")
+
+    val renewalTask = new Runnable() {
+      override def run(): Unit = {
+        updateTokensTask()
+      }
+    }
+    renewalExecutor.schedule(renewalTask, _delay, TimeUnit.MILLISECONDS)
+  }
+
+  /**
+   * Periodic task to login to the KDC and create new delegation tokens. Re-schedules itself
+   * to fetch the next set of tokens when needed.
+   */
+  private def updateTokensTask(): Unit = {
+    try {
+      val freshUGI = doLogin()
+      val creds = obtainTokensAndScheduleRenewal(freshUGI)
+      val tokens = SparkHadoopUtil.get.serialize(creds)
+
+      val driver = driverRef.get()
+      if (driver != null) {
+        logInfo("Updating delegation tokens.")
+        driver.send(UpdateDelegationTokens(tokens))
+      } else {
+        // This shouldn't really happen, since the driver should register way before tokens expire.
+        logWarning("Delegation tokens close to expiration but no driver has registered yet.")
+        SparkHadoopUtil.get.addDelegationTokens(tokens, sparkConf)
+      }
+    } catch {
+      case e: Exception =>
+        val delay = TimeUnit.SECONDS.toMillis(sparkConf.get(CREDENTIALS_RENEWAL_RETRY_WAIT))
+        logWarning(s"Failed to update tokens, will try again in ${UIUtils.formatDuration(delay)}!" +
+          " If this happens too often tasks will fail.", e)
+        scheduleRenewal(delay)
+    }
+  }
+
+  /**
+   * Obtain new delegation tokens from the available providers. Schedules a new task to fetch
+   * new tokens before the new set expires.
+   *
+   * @return Credentials containing the new tokens.
+   */
+  private def obtainTokensAndScheduleRenewal(ugi: UserGroupInformation): Credentials = {
+    ugi.doAs(new PrivilegedExceptionAction[Credentials]() {
+      override def run(): Credentials = {
+        val creds = new Credentials()
+        val nextRenewal = obtainDelegationTokens(creds)
+
+        // Calculate the time when new credentials should be created, based on the configured
+        // ratio.
+        val now = System.currentTimeMillis
+        val ratio = sparkConf.get(CREDENTIALS_RENEWAL_INTERVAL_RATIO)
```

Comment:
why reading this config value every time? (As I have seen other places this is extracted to member vals).

Reply:
What is the advantage to stashing it in a variable? It just decouples the code that is reading the config from the code that is using it.